src/protocol/boolean/comparison.rs

```diff
@@ -2,10 +2,126 @@ use super::or::or;
 use crate::{
     error::Error,
     ff::Field,
-    protocol::{context::Context, BasicProtocols, BitOpStep, RecordId},
+    protocol::{
+        boolean::{random_bits_generator::RandomBitsGenerator, RandomBits},
+        context::Context,
+        BasicProtocols, BitOpStep, RecordId,
+    },
     secret_sharing::Arithmetic as ArithmeticSecretSharing,
 };
 
+// Adapted from 6.1 Interval Test Protocol in "Multiparty Computation for Interval, Equality, and
+// Comparison Without Bit-Decomposition Protocol", Nishide & Ohta, PKC 2007.
+// <https://doi.org/10.1007/978-3-540-71677-8_23>
+//
+// The version in the paper tests c_1 < x < c_2. For us, c_1 is zero (which eliminates the `c < c_1`
+// case enumerated in the paper), we test ≤ rather than <, and we finally return `x > c_2` which is
+// ~(0 ≤ x ≤ c_2).
+//
+// The remainder of this description names the variables consistently with other routines in this
+// file, which map to the paper as follows:
+//
+// Paper  Ours
+// x      a
+// c      b
+// c_1    0
+// c_2    c
+//
+// Goal: Compute `a > c` as `~(0 ≤ a ≤ c)`
+//
+// Strategy:
+//  1. Generate random r
+//  2. Reveal b = a + r
+//  3. Derive bounds r_low and r_high from public values, such that the desired result is a simple
+//     function of `r_low < r` and `r < r_high`.
+//
+// Case 1: b > c
+// b - c - 1 < r < b + 1
```

Comment:
I feel dumb, but I can't figure out why we need to add 1 to $r_{high}$ and subtract 1 from $r_{low}$. I am looking at Fig 5 and the expression and it seems that it is enough for $r < p - c_2 + c$ to be the upper bound for $r$ for $a$ to be outside of ($c_1$, $c_2$). Clearly I am missing something, but can't figure out what. Can someone help me here?

Reply:
It is easy to get confused because of the differences between the paper and the implementation. I think the most succinct argument I have that $\pm 1$ are needed in the Fig. 5 case in the paper is by evaluation at the boundary point $r = p - c_2 + c$. At that point, $a = c - r = c - (p - c_2 + c) = -(p - c_2) \equiv c_2 \pmod p$. Since we are computing the result as $\neg(r_{low} < r < r_{high})$, and we want a negative result when $a = c_2$ (we're going for $c_1 < a < c_2$), we need to adjust $r_{high}$ by $+1$. Similar argument can be made at the low side.

Does that make sense? If not I can offer a few more ways of analyzing it. Also, while going through this, I found a mistake in the correspondence I specified between the paper variables and implementation variables, so thank you for prompting me to spend a few more minutes looking at it so I could catch that.

Reply:
Thanks Andy for your explanation, I think I get it now. Elements in $\mathbb{Z}_p$ are discrete and paper indeed tests for inequality only, hence the shift by 1. In our case we care about equality on the lower bound only, so we keep `-1` and omit `+1`